sunpy/data/manager/manager.py

```diff
@@ -0,0 +1,139 @@
+                    err = KeyError("Hash does not match")
+                    if not details:
+                        # In case we are matching by hash and file does not exist
+                        # That might mean the wrong hash is supplied to decorator
+                        # We match by urls to make sure that is not the case
+                        # XXX: Is this required? Would be useful if users want to
+                        # write functions using `data_manager` or if developers
+                        # are sleepy while writing the code
+                        # TODO: Raise a warning
+                        # https://github.com/sunpy/sunpy/pull/3124#discussion_r291576259
+                        if self._cache_has_file(urls):
+                            raise err
+                        file_path = self._cache.download(urls)
+                    else:
+                        # This is to handle the case when the file is tampered on disk
+                        if hash_file(details['file_path']) != details['file_hash']:
+                            raise err
+                        file_path = details['file_path']
+
+                self._file_cache[name] = file_path
+                return func(*args, **kwargs)
+            return wrapper
+
+        return decorator
+
+    @contextmanager
+    def replace_file(self, name, uri):
+        """
+        Replaces the file by the name with the file provided by the url/path.
+
+        TODO: Hash
+
+        Parameters
+        ----------
+        name: `str`
+            Name of the file provided in the `require` decorator.
+        uri: `str`
+            URI of the file which replaces original file. One of ``http``, ``https``, ``ftp``
+            or ``file``.
+        """
+        try:
+            self._skip_file[name] = uri
+            yield
+        finally:
+            _ = self._skip_file.pop(name, None)
+
+    @contextmanager
+    def skip_hash_check(self):
+        """
+        Disables hash checking temporarily
+
+        Examples
+        --------
+        >>> with remote_data_manager.skip_hash_check():
+        ...     myfunction()
+        """
+        try:
+            self._skip_hash_check = True
+            yield
+        finally:
+            self._skip_hash_check = False
+
+    def get(self, name):
+        """
+        Get the file by name.
+
+        Parameters
+        ----------
+        name: `str`
+            Name of the file given to the data manager, same as the one provided
+            in `~sunpy.data.manager.manager.DataManager.require`.
+
+        Returns
+        -------
+        `pathlib.Path`
+            Path of the file.
+
+        Raises
+        ------
+        `KeyError`
+            If ``name`` is not in the cache.
+        """
+        return pathlib.Path(self._file_cache[name])
+
+    def _cache_has_file(self, urls):
+        for url in urls:
+            if self._cache._get_by_url(url):
+                return True
```

Comment:
This will return `True` if any of the `url` in `urls` is there, but not all. Is that the expected behaviour?

Reply:
Yes. This function is used to match(/confirm) if the file had been downloaded using any of the supplied urls after the matching using hash has failed.

Reply:
Got it! :+1: